### PR TITLE
cheetm/lightfm_deep_dive

### PR DESCRIPTION
**Pull request by cheetm**
### Description

Added LightFM deep dive with accompanying utils

### Related Issues

[#1097](https://github.com/microsoft/recommenders/issues/1097)

### Checklist:

- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.
- [x] This PR is being made to `staging` and not `master`.

### CODE REVIEW

File: notebooks/02_model/README.md [hunk 1]
```diff
@@ -16,5 +16,6 @@ In this directory, notebooks are provided to give a deep dive into training mode
 | [surprise_svd_deep_dive](surprise_svd_deep_dive.ipynb) | Python CPU | Deep dive on a SVD algorithm and implementation.
 | [vowpal_wabbit_deep_dive](vowpal_wabbit_deep_dive.ipynb) | Python CPU | Deep dive into using Vowpal Wabbit for regression and matrix factorization.
 | [fm_deep_dive](fm_deep_dive.ipynb) | Python CPU | Deep dive into factorization machine (FM) and field-aware FM (FFM) algorithm.
+| [lightfm_deep_dive](lightfm_deep_dive.ipynb) | Python CPU | Deep dive into hybrid matrix factorisation model with LightFM.
```
Comment (miguelgfierro) [inline notebooks/02_model/README.md:19]: also I think it would be good to add a similar line in the main readme

File: reco_utils/recommender/lightfm/lightfm_utils.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+from lightfm.evaluation import precision_at_k, recall_at_k, auc_score
+
+def model_perf_plots(df):
+    """
+    Function to plot model performance metrics
+    Args:
+        df (Pandas dataframe): Dataframe in tidy format, with ['epoch','level','value'] columns
+    Returns:
+        matplotlib axes
+    """
```
Comment (miguelgfierro) [inline reco_utils/recommender/lightfm/lightfm_utils.py:16]: a minor comment for consistency with the rest of the library, we would do the comments like this:
```
    """Function to plot model performance metrics

    Args:
        df (pd.DataFrame): Dataframe in tidy format, with ['epoch','level','value'] columns

    Returns:
        obj: matplotlib axes
    """
```

Something that we (painfully) realised some time ago is that if we don't add a new line between Returns: and the previous comment, the docstrings are not parsed by sphinx, so we can't build the documentation

File: reco_utils/recommender/lightfm/lightfm_utils.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    # fit model and store train/test metrics at each epoch 
+    for epoch in range(no_epochs):
+    #     print(f'Epoch: {epoch}/{epochs}')
+        model.fit_partial(interactions=train_interactions, epochs=1,
+                                   num_threads=no_threads, **kwargs)
+        model_prec_train[epoch] = precision_at_k(model, train_interactions, k=k, **kwargs).mean()
+        model_prec_test[epoch] = precision_at_k(model, test_interactions, k=k, **kwargs).mean()
+
+        model_rec_train[epoch] = recall_at_k(model, train_interactions, k=k, **kwargs).mean()
+        model_rec_test[epoch] = recall_at_k(model, test_interactions, k=k, **kwargs).mean()
```
Comment (miguelgfierro) [inline reco_utils/recommender/lightfm/lightfm_utils.py:76]: please take a look at the previous comments in the notebook

File: notebooks/02_model/lightfm_deep_dive.ipynb [hunk 4]
```diff
@@ -0,0 +1,1358 @@
+{
```
Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>can you please make sure that these formulas are rendered correctly on Github:</p><pre class="ql-syntax" spellcheck="false">\begin{equation}&nbsp;q_u = \sum_{j \in f_u} e_{j}^{U}&nbsp;\end{equation}

\begin{equation}&nbsp;p_i = \sum_{j \in f_i} e_{j}^{I}&nbsp;\end{equation}

Similarly the biases for user&nbsp;

\begin{equation}&nbsp;b_u = \sum_{j \in f_u} b_{j}^{U}&nbsp;\end{equation}

\begin{equation}&nbsp;b_i = \sum_{j \in f_i} b_{j}^{I}&nbsp;\end{equation}
</pre><p>I don't see them right</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>would it be possible to have all the imports on the first cell and then the parameters on a different cell? We are doing that in the rest of the notebook. Also we use the parameters cell to execute the notebook tests, please see more details here: <a href="https://github.com/microsoft/recommenders/blob/master/tests/README.md#how-to-create-tests-on-notebooks-with-papermill" target="_blank">https://github.com/microsoft/recommenders/blob/master/tests/README.md#how-to-create-tests-on-notebooks-with-papermill</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>Here two comments:</p><p>1) We follow the single responsibility pattern <a href="https://github.com/Microsoft/Recommenders/wiki/Coding-Guidelines#single-responsibility" target="_blank">https://github.com/Microsoft/Recommenders/wiki/Coding-Guidelines#single-responsibility</a>, as well as explicit is better than implicit. This effictively means that we normally have one part called fit, that trains the model and another one called predict or recommend that predicts on the test set. This can be an internal method, or if the library already has a fit, use it directly.</p><p>2) The plot shows a high overfitting, are you sure that this <code>model.fit_partial(interactions=train_interactions, epochs=1</code>  works right?</p><p>3) for consistency it would be good to use our own precision_at_k and other metrics. One example that would be very similar to your is <a href="https://github.com/microsoft/recommenders/blob/master/notebooks/02_model/surprise_svd_deep_dive.ipynb" target="_blank">https://github.com/microsoft/recommenders/blob/master/notebooks/02_model/surprise_svd_deep_dive.ipynb</a>. There we are using a similar approach, using an external library, called surprise, and try to add as less code as needed to show how it works</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>I like how you structured the notebook, it is clear the different modes</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>this is super useful, a lot of people were asking for this</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='58'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>could you please add the years in 1 and 4?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

File: reco_utils/recommender/lightfm/lightfm_utils.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
+import pandas as pd
```
Comment (miguelgfierro) [inline reco_utils/recommender/lightfm/lightfm_utils.py:1]: could you please add the dependency into generate_conda_file?

**Comment by miguelgfierro** [comment]
@cheetm this looks really good, I added some comments. It would be great if you could connect with Nile, she has already a lot of experience on the repo

**Comment by miguelgfierro** [comment]
@niwilso I'm trying to add you as reviewer but for some reason I can't. Feel free to add comments and suggestions as well

File: notebooks/02_model/lightfm_deep_dive.ipynb (hunk 4 above)
Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: Thanks!

Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: fixed

Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: I've added explanation why fit_partial was used, also added example re how to use fit method and compared both models

Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: done, also added parameters cell.

File: notebooks/02_model/lightfm_deep_dive.ipynb [hunk 6]
```diff
@@ -0,0 +1,1505 @@
...
```
Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>just a reminder to remove this content when the notebook is finished</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>in other notebooks what we have is a step where we perform the predicion and then another step/cell where we compute the metrics (precision_at_k, recall, etc.) </p><p> </p><p>Then it would be great that on a different cell we add the comparison between fit and fit_partial </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (miguelgfierro) [inline notebooks/02_model/lightfm_deep_dive.ipynb:1]: <p>copying same comment here:</p><p>for consistency it would be good to use our own precision_at_k and other metrics. One example that would be very similar to your is&nbsp;<a href="https://github.com/microsoft/recommenders/blob/master/notebooks/02_model/surprise_svd_deep_dive.ipynb" target="_blank">https://github.com/microsoft/recommenders/blob/master/notebooks/02_model/surprise_svd_deep_dive.ipynb</a>. There we are using a similar approach, using an external library, called surprise, and try to add as less code as needed to show how it works</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1096/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: Hi @miguelgfierro , thanks for your comments. I agree with you on this point using the existing methods to extract metrics. 
The difficulty being LightFM doesn't predict the user-item ratings directly (the prediction scores are only for ranking purposes), hence it would involve rather significant modification to get it to work with the existing codes.
Let me know how do you want to progress this/if you still think this is critical?
Thanks

Comment (miguelgfierro) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: there has to be a mapping between their ranking metrics and ours. 

It is in fact important, the reason is because that enable us to compare between algos. But obviously, it depends whether you have bandwidth or not. In case, you don't have bandwidth, I would prefer to not show the metrics, so we don't cause confusion

Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: Hi there, a bit puzzled in how to get built-in evaluation metrics to work - if I create a sample test df with all predictions = 1 (for argument sage), I get a 0.886 Prec@K - see following: 

Screenshot from 2020-05-05 19-26-24

However, if I assign the prediction column to random ints, I also get exactly the same Prec@K - see attached:

Screenshot from 2020-05-05 19-30-53

In fact have similar issues with NDCG and Recall@K as well, so not exactly sure whether I am using these methods correctly. Would be grateful if you could advise.

Thanks

Comment (miguelgfierro) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: @yueguoguo worked on the metrics, he might be able to help

Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: Thanks @miguelgfierro, I've solved that and I'll try to re-structure the deep dive notebook differently later.

Comment (miguelgfierro) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: awesome! I'll close this item then

File: notebooks/02_model/lightfm_deep_dive.ipynb (hunk 4 above)
Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: I have changed and restructured the notebook, now it contains steps to map the data into a format suitable for the repo's evaluation metrics.

File: notebooks/02_model/lightfm_deep_dive.ipynb (hunk 6 above)
Comment (cheetm) [reply notebooks/02_model/lightfm_deep_dive.ipynb:1]: Thanks, now it contains sections on how to get the data into the right format for the repo's metrics.

**Comment by anargyri** [comment]
Hey @cheetm this PR is very nice. Thanks for the great work! 
My main comment is that we follow the practice of using the metrics computations from this repo, instead of the ones from `lightfm.evaluation`. This allows for consistency and comparisons between different algorithms and frameworks.

**Comment by cheetm** [comment]
> Hey @cheetm this PR is very nice. Thanks for the great work!
> My main comment is that we follow the practice of using the metrics computations from this repo, instead of the ones from `lightfm.evaluation`. This allows for consistency and comparisons between different algorithms and frameworks.

Hi @anargyri thanks for your comments.
Perhaps it is not clear from the notebook, all the evaluation metrics were computed using the repo's methods up to section 3. In section 4 the progress were tracked and measured using LightFM's method because they are significantly more efficient.

You can use the output from cell 18 and/or cell 32 for comparison with other algorithms.

Hope thats okay.
Thanks

File: reco_utils/recommender/lightfm/lightfm_utils.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+import numpy as np
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+import lightfm
+from lightfm.evaluation import precision_at_k, recall_at_k, auc_score
```
Comment (anargyri) [inline reco_utils/recommender/lightfm/lightfm_utils.py:7]: You are using the `lightfm.evaluation` methods in this file. I think it would be better, for longer term maintenance of the code, to replace them with the ones from this repo. 
Also `auc_score` is not used here and can be removed.

**Comment by miguelgfierro** [comment]
@cheetm about this:
`In section 4 the progress were tracked and measured using LightFM's method because they are significantly more efficient.`

- do you have some metrics of the difference?
- do you know where the speed up comes from and/or how to improve our current metrics?

FYI @yueguoguo

File: reco_utils/recommender/lightfm/lightfm_utils.py (hunk 7 above)
Comment (cheetm) [reply reco_utils/recommender/lightfm/lightfm_utils.py]: Thanks and removed.
the precision_at_k and recall_at_k are for section 4 as explained above, i'll keep them in else section 4 will takes forever to execute.

**Comment by cheetm** [comment]
> @cheetm about this:
> `In section 4 the progress were tracked and measured using LightFM's method because they are significantly more efficient.`
> 
> * do you have some metrics of the difference?
> * do you know where the speed up comes from and/or how to improve our current metrics?
> 
> FYI @yueguoguo

I have added section 3.6  to demonstrate the difference (albeit very naively), the difference is so great I don't think we need to resort to timeit. 
Also note that if I were to use the repo's metrics, each epoch in section 4 would require ~13 minutes to compute which is why I've decided to use the LightFM's methods.

I haven't got time to properly investigate your second point, but I am guessing if you have already fitted the latent matrices, the scores can be retrieved much more efficiently all at once instead of individually.
